### ISSUE TITLE
logstash refuses to stop and exit, gets stuck

### ISSUE THREAD

**Issue by FlorinAndrei** [elastic/logstash#4703]
logstash-2.2.2, elasticsearch-1.7.5, java-1.8.0_72-b15, ubuntu-14.04 64bit, couchdb-1.6.1

```
input {
  couchdb_changes {
    always_reconnect => true
    db => "vendnext_new_no_auth"
    heartbeat => 5000
    host => "couch31-prod-uswest2-aws.domain"
    port => 5984
    reconnect_delay => 10
  }
}
filter {
  mutate {
    remove_field => [ "stripe", "moneris", "roles", "_rev", "freeVendUses", "surcharge", "debugMessages" ]
    convert => [ "[doc][debugMessages]", "string"]
  }
}
output {
  elasticsearch {
    action => "index"
    hosts => "localhost:9200"
    index => "vendnext_new_no_auth_index"
    manage_template => false
    script => "domain-filter"
    script_lang => "javascript"
    script_type => "file"
  }
}
```

Overall it appears to work just fine. But the problems begin when I try to stop it.

"service logstash stop" stops the updates, but the logstash process does not quit. Repeating "service logstash stop" makes no difference. Logstash appears to be frozen - no updates sent to ElasticSearch, but it does not quit either.

After that, "service logstash status" reports the service as still running.

"kill $PID" doesn't kill it. The only way to kill it is "kill -9 $PID".

**Comment by ebuildy** [comment]
Yes LogStash is kept alive until there is no more event to process and sometimes (often ;-( ) never stops.

As you can read at https://www.elastic.co/blog/logstash-2-1-0-released:


> in Logstash 2.0.0, we've added the ability to start Logstash with a CLI option (--allow-unsafe-shutdown) that allows it to predictably shutdown when the operator initiates it. This is particularly useful in situations where the processing pipeline is blocked, trying to communicate to external sources. Users were frustrated with not being able to shutdown Logstash in these situation. Please be aware that shutting down Logstash could still lead to loss of any in-flight messages.

**Comment by jordansissel** [comment]
@FlorinAndrei hmm, I haven't tried to reproduce this, but my hypothesis is that the couchdb_changes input is failing to terminate somehow. On shutdown, logstash will tell all inputs to close and wait until all events have left the pipeline before terminating -- this is why my idea is maybe a bug in couchdb_changes causing it to not close when asked.

**Comment by ebuildy** [comment]
You can dump Java threads to see what's going on:

    jstack PID

For instance, elasticsearch output waiting:

````
"Ruby-0-Thread-4: /opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-output-elasticsearch-2.5.1-java/lib/logstash/outputs/elasticsearch/buffer.rb:78" #16 daemon prio=5 os_prio=0 tid=0x00007f8ce8186000 nid=0x20 in Object.wait() [0x00007f8d1c32b000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	at org.jruby.RubyThread.sleep(RubyThread.java:1002)
	- eliminated <0x00000005d4597240> (a org.jruby.RubyThread)
	- locked <0x00000005d4597240> (a org.jruby.RubyThread)
	at org.jruby.RubyKernel.sleep(RubyKernel.java:803)
	at org.jruby.RubyKernel$INVOKER$s$0$1$sleep.call(RubyKernel$INVOKER$s$0$1$sleep.gen)
	at org.jruby.internal.runtime.methods.JavaMethod$JavaMethodN.call(JavaMethod.java:667)
	at org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:206)
	at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:168)
	at org.jruby.ast.FCallOneArgNode.interpret(FCallOneArgNode.java:36)
	at org.jruby.ast.NewlineNode.interpret(NewlineNode.java:105)
	at org.jruby.ast.BlockNode.interpret(BlockNode.java:71)
	at org.jruby.evaluator.ASTInterpreter.INTERPRET_BLOCK(ASTInterpreter.java:112)
	at org.jruby.runtime.Interpreted19Block.evalBlockBody(Interpreted19Block.java:206)
	at org.jruby.runtime.Interpreted19Block.yield(Interpreted19Block.java:157)
	at org.jruby.runtime.Interpreted19Block.yieldSpecific(Interpreted19Block.java:130)
	at org.jruby.runtime.Block.yieldSpecific(Block.java:111)
	at org.jruby.RubyKernel.loop(RubyKernel.java:1479)
	at org.jruby.RubyKernel$INVOKER$s$0$0$loop.call(RubyKernel$INVOKER$s$0$0$loop.gen)
	at org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:316)
	at org.jruby.runtime.callsite.CachingCallSite.callBlock(CachingCallSite.java:145)
	at org.jruby.runtime.callsite.CachingCallSite.callIter(CachingCallSite.java:154)
	at org.jruby.ast.FCallNoArgBlockNode.interpret(FCallNoArgBlockNode.java:32)
	at org.jruby.ast.NewlineNode.interpret(NewlineNode.java:105)
	at org.jruby.evaluator.ASTInterpreter.INTERPRET_BLOCK(ASTInterpreter.java:112)
	at org.jruby.runtime.Interpreted19Block.evalBlockBody(Interpreted19Block.java:206)
	at org.jruby.runtime.Interpreted19Block.yield(Interpreted19Block.java:194)
	at org.jruby.runtime.Interpreted19Block.call(Interpreted19Block.java:125)
	at org.jruby.runtime.Block.call(Block.java:101)
	at org.jruby.RubyProc.call(RubyProc.java:300)
	at org.jruby.RubyProc.call(RubyProc.java:230)
	at org.jruby.internal.runtime.RubyRunnable.run(RubyRunnable.java:99)
	at java.lang.Thread.run(Thread.java:745)
````

**Comment by andrewvc** [comment]
@FlorinAndrei I'll just echo what @ebuildy said, you should also see 'STALLING PIPELINE' messages in your logs if its stuck shutting down.

You can always force it to shutdown with `--allow-unsafe-shutdown` as @ebuildy said.

**Comment by jordansissel** [comment]
@FlorinAndrei ping? I think we're waiting on some information from you :)

**Comment by FlorinAndrei** [comment]
Sorry, I was busy and lost track of this issue.

I deleted the ElasticSearch index, and re-created it again, empty. I started Logstash, importing data from CouchDB. I let it absorb data for a minute or two. Then I tried to shut down the logstash service. It got stuck as usual.

Here's the jstack output:

https://gist.github.com/FlorinAndrei/84992a422829e15c294c

Here's logstash.log:

https://gist.github.com/FlorinAndrei/6bf1b43f01d6b54d7d0c

Here's logstash.stdout:

https://gist.github.com/FlorinAndrei/8d5c243db39b970a456a

There's nothing in logstash.err

The log files were empty before I did this experiment.

**Comment by Dr-Steve** [comment]
On windows using git-bash sending a ctrl-c or a kill -9 doesn't even work.  I literally have to restart my machine whenever logstash gets stuck while testing.  Would be nice to have an easier way to kill it.

**Comment by yraedry** [comment]
I use Iobit Unloker in the data folde rto unlock the process and when i stop the proccess, I delete the data folder

**Comment by hsavasere** [comment]
Facing the same issue in Ubuntu 20.04 as well after running the command 
_**sudo apt-get remove --purge logstash**_

**Comment by SupremeGit** [comment]
This is an absolute disgrace.
When a product cannot be reliably and easily stopped, it is not fit to be used by anyone.

Whoever the developers of logstash are, they need to stop insulting their users by allowing this problem to continue.

Absolutely disgusting.

**Comment by andsel** [comment]
Thanks @crockk for your feedback,
the original issue regards Logstash 2.2.2 which is not anymore supported. Did you experience the problem with a more recent Logstash version?

**Comment by j0nny55555** [comment]
Yes, this still appears to be happening. While I have some experience with ELK, I must admit, I don't think I could give you output files like @FlorinAndrei did. Compliments Florin, I'd love to have the method/roll out you used to collect that data.

I do have an extra plug-in installed. A Logstash opnsense plug-in.

Is there anything you could point me to to help? I am using the latest packages. @andsel

**Comment by FlorinAndrei** [comment]
@j0nny55555 sorry, that was a long time ago and I don't remember what I did back then to troubleshoot it.

**Comment by andsel** [comment]
@j0nny55555 would be nice to create a reproducer. Could you share the config you used, at least a minimal reproducing pipeline, specifying the versions of Logstash and plugins involved in the pipeline, please?

**Comment by j0nny55555** [comment]
@andsel and @FlorinAndrei thank you both so much for your response. i've only begun to be active here, and it is only on the side - speaking to availability.

in about two weeks time i might have a full rebuild of my logstash host and i've also got filebeat in the mix too. thank you again, i'll update to success/failure.
